src/server/main_service.cc

```diff
@@ -879,35 +929,85 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
   }
 
-  cntx->conn_state.exec_state = ConnectionState::EXEC_INACTIVE;
-  cntx->conn_state.exec_body.clear();
-
+  UnwatchAllKeys(cntx);
+  cntx->conn_state.exec_info.Clear();
   rb->SendOk();
 }
 
+// Return true if non of the connections watched keys expired.
+bool CheckWatchedKeyExpiry(ConnectionContext* cntx, const CommandRegistry& registry) {
+  static char EXISTS[] = "EXISTS";
+  auto& exec_info = cntx->conn_state.exec_info;
+
+  CmdArgVec str_list(exec_info.watched_keys.size()+1);
+  str_list[0] = MutableSlice{EXISTS, strlen(EXISTS)};
+  for (size_t i = 1;  i < str_list.size(); i++) {
+    string& s = exec_info.watched_keys[i-1];
+    str_list[i] = MutableSlice{s.data(), s.size()};
+  }
+
+  atomic_uint32_t watch_exist_count{0};
+  auto cb = [&watch_exist_count, &exec_info](Transaction* t, EngineShard* shard) {
+    ArgSlice args = t->ShardArgsInShard(shard->shard_id());
+    auto res = GenericFamily::OpExists(t->GetOpArgs(shard), args);
+    watch_exist_count.fetch_add(res.value_or(0), memory_order_relaxed);
+    return OpStatus::OK;
+  };
+
+  VLOG(1) << "Checking expired watch keys";
+
+  cntx->transaction->SetExecCmd(registry.Find(EXISTS));
+  cntx->transaction->InitByArgs(cntx->conn_state.db_index,
+    CmdArgList{str_list.data(), str_list.size()});
+  OpStatus status = cntx->transaction->ScheduleSingleHop(std::move(cb));
+  CHECK_EQ(OpStatus::OK, status);
+
+  // The comparison can still be true even if a key expired due to another one being created.
+  // So we have to check the watched_dirty flag, which is set if a key expired.
+  return watch_exist_count.load() == exec_info.watched_existed
+    && !exec_info.watched_dirty.load();
+}
+
 void Service::Exec(CmdArgList args, ConnectionContext* cntx) {
   RedisReplyBuilder* rb = (*cntx).operator->();
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_INACTIVE) {
+  if (!cntx->conn_state.exec_info.IsActive()) {
     return rb->SendError("EXEC without MULTI");
   }
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_ERROR) {
-    cntx->conn_state.exec_state = ConnectionState::EXEC_INACTIVE;
-    cntx->conn_state.exec_body.clear();
+  auto& exec_info = cntx->conn_state.exec_info;
+
+  if (exec_info.state == ConnectionState::ExecInfo::EXEC_ERROR) {
+    UnwatchAllKeys(cntx);
+    exec_info.Clear();
     return rb->SendError("-EXECABORT Transaction discarded because of previous errors");
   }
 
-  VLOG(1) << "StartExec " << cntx->conn_state.exec_body.size();
-  rb->StartArray(cntx->conn_state.exec_body.size());
-  if (!cntx->conn_state.exec_body.empty()) {
+  if (exec_info.watched_dirty.load()) {
+    UnwatchAllKeys(cntx);
+    exec_info.Clear();
+    return rb->SendNull();
+  }
+
+  // EXEC should not run if any of the watched keys expired.
+  if (!exec_info.watched_keys.empty() && !CheckWatchedKeyExpiry(cntx, registry_)) {
+    cntx->transaction->UnlockMulti();
+    UnwatchAllKeys(cntx);
+    exec_info.Clear();
+    return rb->SendNull();
+  }
```

Comment:
Again, this if has a lot to worry about if it passes

Reply:
why do not understand what you mean. Also, should not you send an error if something transaction was aborted?

Reply:
No, just NULL. See https://redis.io/docs/manual/transactions/ under optimistic-locking

> I do not understand what you mean

Instead of repeating the same code 4 times, we could use a defer-like (like in golang but with a destructor) construction to list in once. I just though of this when I had forgotten to put in into one of the code-paths and found it when testing.

Reply:
`absl::Cleanup` is what you want.